### PR TITLE
Fy formats 

### LINKED ISSUE #129: Create functions to convert between Financial Year types
We mostly use FY in the form of 1718 (2017/18) but sometimes we also need it as 2017. 

Make functions that swap between the two formats.

### PR DESCRIPTION
**Pull request by catherineholland1**
closes #129 
Functions created to convert normal financial year form into the alternate form and the reverse. 

i.e 1718 to 2017 and the reverse

### CODE REVIEW

File: Source Linkage Files.Rproj [hunk 1]
```diff
@@ -0,0 +1,13 @@
+Version: 1.0
```
Comment (Moohan) [inline Source Linkage Files.Rproj:1]: I'm not sure why this file was created. Could you just delete it please? or revert the commit which added it (would be better but slightly more work!)

File: R/convert_year_types.R [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+#' @param year vector of years in the form '2017'
+#'
+#' @return a vector of years in the normal form '1718'
+#' @export
+#'
+#' @examples
+#' convert_year_to_fyyear(c("1718", "1819"))
+convert_year_to_fyyear <- function(year) {
+  fyyear <- paste0(substr(year, 3, 4), as.numeric(substr(year, 3, 4)) + 1)
+  return(as.numeric(fyyear))
```
Comment (Moohan) [inline R/convert_year_types.R:29]: I think we probably want to return a character?

At least it should be consistent i.e. both functions take and return the same type (either integer or character)

File: tests/testthat/test-convert_year_types.R [hunk 3]
```diff
@@ -0,0 +1,61 @@
+test_that("Can convert a year from normal financial year to the alternate form", {
```
Comment (Moohan) [inline tests/testthat/test-convert_year_types.R:1]: Tests are good.
No need to test all years, one or two is fine. Testing edge cases is good though i.e. earliest file = 2011/12.  Also, test incorrect inputs e.g. what if we run `convert_fyyear_to_year("2017)`?

**Review by Moohan — CHANGES_REQUESTED** [review]
Looks good. There's a couple of changes suggested :)

File: Source Linkage Files.Rproj (hunk 1 above)
Comment (catherineholland1) [reply Source Linkage Files.Rproj]: commit reverted

File: R/convert_year_types.R (hunk 2 above)
Comment (catherineholland1) [reply R/convert_year_types.R:29]: removed as.numeric() from function to return a character variable instead

File: tests/testthat/test-convert_year_types.R (hunk 3 above)
Comment (catherineholland1) [reply tests/testthat/test-convert_year_types.R:1]: removed unnecessary year tests

Would I include an error message to make sure that the correct form has been inputted?

**Comment by Moohan** [comment]
/style

Just testing this GH action ^^ in theory it should run styler and re-commit 😉

**Review by catherineholland1 — COMMENTED** [review]
all changes complete

File: R/convert_year_types.R [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+#'
+#' @examples
+#' fyyears <- c("1718", "1819")
+#' convert_fyyear_to_year(fyyears)
+#' [1] "2017" "2018"
+convert_fyyear_to_year <- function(fyyear) {
+  for (i in 1:length(fyyear)) {
+    if (substr(fyyear[i], 1, 2) > substr(fyyear[i], 3, 4)) {
+      stop ("Year has been entered in the wrong format, try again using form `1718` or use function `convert_year_to_fyyear` to convert to the financial year form.")
+  }}
```
Comment (Moohan) [inline R/convert_year_types.R:18]: You can get rid of the whole loop construct with `any()` i.e. if any of the supplied values are bad throw an error.
```suggestion
    if (any(substr(fyyear[i], 1, 2) > substr(fyyear[i], 3, 4))) {
      stop("Year has been entered in the wrong format, try again using form `1718` or use function `convert_year_to_fyyear` to convert to the financial year form.")
  }
```

We could also make the check more strict than just greater than, i.e. second part must be first part + 1:
```suggestion
  first_part <- as.integer(substr(fyyear, 1, 2))
  second_part <- as.integer(substr(fyyear, 3, 4))
  if (any(first_part != second_part + 1L)) {
    stop("Year has been entered in the wrong format, try again using form `1718` or use function `convert_year_to_fyyear` to convert to the financial year form.")
  }
```

File: R/convert_year_types.R [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+#'
+#' @examples
+#' years <- c("2017", "2018")
+#' convert_year_to_fyyear(years)
+#' [1] "1718" "1819"
+convert_year_to_fyyear <- function(year) {
+  for (i in 1:length(year)){
+    if (substr(year[i], 1, 2) != "20") {
+      stop ("Year has been entered in the wrong format, try again using form `2017` or use function `convert_fyyear_to_year` to convert to alternate year form.")
+  }}
```
Comment (Moohan) [inline R/convert_year_types.R:43]: Replace the loop with `any()`

File: R/convert_year_types.R [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+#'
+#' @param fyyear vector of financial years in the form '1718'
+#'
+#' @return a vector of years in the alternate form '2017'
+#' @export
+#'
+#' @examples
+#' fyyears <- c("1718", "1819")
+#' convert_fyyear_to_year(fyyears)
+#' [1] "2017" "2018"
```
Comment (Moohan) [inline R/convert_year_types.R:13]: Remove this line please, the examples are run and the output is generated when you run `devtools::document`. So this isn't needed (and possibly breaks the documentation output)

File: R/convert_year_types.R [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+#'
+#' @param year vector of years in the form '2017'
+#'
+#' @return a vector of years in the normal financial year form '1718'
+#' @export
+#'
+#' @examples
+#' years <- c("2017", "2018")
+#' convert_year_to_fyyear(years)
+#' [1] "1718" "1819"
```
Comment (Moohan) [inline R/convert_year_types.R:38]: Remove this line as above, please.

**Review by Moohan — CHANGES_REQUESTED** [review]
Almost there, a couple more changes though please:

- [ ] Replace loops with use of `any()`
- [ ] Remove 'output' lines from documentation comments
- [ ] Add a test to check for the error e.g. `expect_error(convert_fyyear_to_year("2017"), "Year has been entered in the wrong format")`
- [ ] Run [styler](https://styler.r-lib.org/index.html) - Just install the package then use the addin 😉

File: R/convert_year_types.R (hunk 7 above)
Comment (catherineholland1) [reply R/convert_year_types.R]: Just the [1] line or calling the function as well?

File: R/convert_year_types.R (hunk 4 above)
Comment (catherineholland1) [reply R/convert_year_types.R]: When I include your above code the test fails? but unsure how as all of the test years second part are +1 from the first

File: R/convert_year_types.R (hunk 7 above)
Comment (Moohan) [reply R/convert_year_types.R]: Yep just the 'output' `[1]` line as you've done 👍

File: R/convert_year_types.R (hunk 4 above)
Comment (Moohan) [reply R/convert_year_types.R]: Hmm, I'll have a quick look but I'm happy with it how it is too, I'll either push a change or approve shortly!

**Review by Moohan — APPROVED** [review]

**Comment by Moohan** [comment]
All changes approved. @catherineholland1 I'll leave this to you to merge it 😉 I would usually do a 'squash and merge' (see the drop-down arrow) for this type of change as ultimately it's quite a simple addition. Squash and merge will combine all the commits into one (it combines all the messages too and gives you chance to edit it).
